components/Visualization/ClusterForceLayout/index.js

```diff
@@ -0,0 +1,302 @@
+        width: PropTypes.number,
+        height: PropTypes.number,
+    }).isRequired,
+    data: PropTypes.arrayOf(PropTypes.shape({
+        value: PropTypes.string,
+        score: PropTypes.number,
+    })),
+    selectedData: PropTypes.shape({
+        value: PropTypes.string,
+        score: PropTypes.number,
+    }),
+    idSelector: PropTypes.func.isRequired,
+    groupSelector: PropTypes.func,
+    valueSelector: PropTypes.func,
+    scaleFactor: PropTypes.number,
+    onHover: PropTypes.func,
+    className: PropTypes.string,
+    colorScheme: PropTypes.func,
+    margins: PropTypes.shape({
+        top: PropTypes.number,
+        right: PropTypes.number,
+        bottom: PropTypes.number,
+        left: PropTypes.number,
+    }),
+};
+const defaultProps = {
+    data: {
+    },
+    margins: {
+        top: 0,
+        right: 0,
+        bottom: 0,
+        left: 0,
+    },
+    selectedData: {
+        score: 10,
+        value: 'gold',
+        cluster: 0,
+    },
+    groupSelector: d => d.cluster,
+    valueSelector: d => d.score,
+    scaleFactor: 2,
+    onHover: () => true,
+    className: '',
+    colorScheme: interpolateRainbow,
+};
+
+class ClusterForceLayout extends PureComponent {
+    static propTypes = propTypes;
+    static defaultProps = defaultProps;
+
+    componentDidMount() {
+        this.drawChart();
+    }
+    componentWillReceiveProps(nextProps) {
+        if (nextProps.data !== this.props.data) {
+            this.updateData(nextProps);
+        }
+    }
+
+    componentDidUpdate() {
+        this.redrawChart();
+    }
+
+    onMouseOver = (d) => {
+        this.props.onHover(d);
+        select(this.style)
+            .style('display', 'block');
+    }
+
+    onMouseMove = (d) => {
+        select(this.tooltip)
+            .html(`<span class=${styles.id}>${d.id}</span>
+            <span class=${styles.score}>${d.radius}</span>`)
+            .style('display', 'block')
+            .style('top', `${event.pageY - 50}px`)
+            .style('left', `${event.pageX - 30}px`);
+    }
+
+    onMouseOut = () => {
+        select(this.tooltip)
+            .style('display', 'none');
+    }
+
+    getMaxNode = (data, valueSelector) => data.reduce((previous, current) => (
+        valueSelector(current) > valueSelector(previous) ? current : previous
+    ))
+
+    dragstarted = (d) => {
+        if (!event.active) this.simulation.alphaTarget(0.3).restart();
+        d.fx = d.x; // eslint-disable-line no-param-reassign
+        d.fy = d.y; // eslint-disable-line no-param-reassign
+    }
+
+    dragged = (d) => {
+        d.fx = event.x; // eslint-disable-line no-param-reassign
+        d.fy = event.y; // eslint-disable-line no-param-reassign
+    }
+
+    dragended = (d) => {
+        if (!event.active) this.simulation.alphaTarget(0);
+        d.fx = null; // eslint-disable-line no-param-reassign
+        d.fy = null; // eslint-disable-line no-param-reassign
+    }
+
+    save = () => {
```

Comment:
save should not be called using ref

Reply:
the save function is not called with ref.

Reply:
nice